File: 01_original_game/pong.js
const BLACK = "#000000";
const RED = "#ff0000";
const BLUE = "#0000ff";
const WHITE = "#ffffff";

let ball = {
    xpos: 395,
    ypos: 295,
    xspeed: 0,
    yspeed: 0,
}

const leftPaddle = {
    xpos: 25,
    ypos: 300
}

const rightPaddle = {
    xpos: 750,
    ypos: 300
}

const initialPosition = { ...ball };

const MOVE = {
    STATIONARY: 0,
    UP: 1,
    DOWN: 2
}

const timings = new Array(1000);

let move = MOVE.STATIONARY;

let fps = 0;
let frames = 0;
let totalFrames = 0;
let time = Date.now();

let leftScore = 0;
let rightScore = 0;

window.addEventListener("keydown", function (event) {
    if (event.key === "ArrowDown") {
        move = MOVE.DOWN;
    }
    if (event.key === "ArrowUp") {
        move = MOVE.UP;
    }
});

window.addEventListener("keyup", function (event) {
    move = MOVE.STATIONARY;
});


function calculateFps() {
    frames += 1;
    totalFrames += 1;
    if (Date.now() - time > 1000) {
        fps = frames / ((Date.now() - time) / 1000)
        frames = 0;
        time = Date.now();
    }
}

function getAIMove() {

    let idealPosition;

    if (ball.xspeed > 0) {
        idealPosition = ball.ypos;
    } else {
        turns = (ball.xpos - 50) / (-1 * ball.xspeed);
        idealPosition = ball.ypos + (ball.yspeed * turns);
    }

    if (idealPosition < leftPaddle.ypos) {
        return MOVE.UP;
    }
    if (idealPosition > leftPaddle.ypos) {
        return MOVE.DOWN;
    }
    return MOVE.STATIONARY;
}

function ballCollidesWithTopOrBottom() {
    return ball.ypos - 5 < 0 || ball.ypos + 5 > 595;
}

function ballCollidesWithLeftPaddle() {
    return (
        ball.xpos - 5 < leftPaddle.xpos + 5 &&
        ball.ypos - 5 < leftPaddle.ypos + 50 &&
        ball.ypos + 5 > leftPaddle.ypos - 50
    );
}

function ballCollidesWithRightPaddle() {
    return (
        ball.xpos + 5 > rightPaddle.xpos &&
        ball.ypos - 5 < rightPaddle.ypos + 50 &&
        ball.ypos + 5 > rightPaddle.ypos - 50
    );
}

function ballCollidesWithPaddle() {
    return ballCollidesWithLeftPaddle() || ballCollidesWithRightPaddle();
}

function calculateReflectionFactor() {
    let paddle = leftPaddle;
    if (ball.xpos > 400) {
        paddle = rightPaddle;
    };

    const factor = ball.ypos - paddle.ypos;
    return factor / 100;
}

function ballScoreOnRight() {
    return ball.xpos > 775;
}

function ballScoreOnLeft() {
    return ball.xpos < 25;
}

function update() {
    if (ball.xspeed === 0 && ball.yspeed === 0) {
        ball.xspeed = 1;
        ball.yspeed = (Math.random() * 2) - 1;
    }

    if (move === MOVE.UP && rightPaddle.ypos > 50) {
        rightPaddle.ypos -= 1;
    }
    if (move === MOVE.DOWN && rightPaddle.ypos < 550) {
        rightPaddle.ypos += 1;
    }

    const AIMove = getAIMove();
    console.log(AIMove);
    if (AIMove === MOVE.UP && leftPaddle.ypos > 50) {
        leftPaddle.ypos -= 1;
    }
    if (AIMove === MOVE.DOWN && leftPaddle.ypos < 550) {
        leftPaddle.ypos += 1;
    }

    // ball touch on sides and paddles
    if (ballCollidesWithTopOrBottom()) {
        ball.yspeed = -ball.yspeed;
    }
    if (ballCollidesWithPaddle()) {
        ball.xspeed = -ball.xspeed;
        ball.xspeed *= 1.05;
        ball.yspeed += calculateReflectionFactor();
    }

    // create score
    if (ballScoreOnRight()) {
        ball = { ...initialPosition };
        leftScore += 1;
    }
    if (ballScoreOnLeft()) {
        ball = { ...initialPosition };
        rightScore += 1;
    }

    ball.xpos += ball.xspeed;
    ball.ypos += ball.yspeed;

}


function drawBlackBackground(ctx) {
    ctx.fillStyle = BLACK;
    ctx.fillRect(0, 0, 800, 600);
}

function drawDottedLine(ctx) {
    ctx.strokeStyle = WHITE;
    ctx.beginPath();
    ctx.setLineDash([5, 15]);
    ctx.moveTo(400, 0);
    ctx.lineTo(400, 600);
    ctx.lineWidth = 10;
    ctx.stroke();

    // reset line
    ctx.setLineDash([]);
    ctx.lineWidth = 1;
}

function drawPaddles(ctx) {
    ctx.fillStyle = RED;
    ctx.fillRect(leftPaddle.xpos, leftPaddle.ypos - 50, 25, 100);
    ctx.fillStyle = BLUE;
    ctx.fillRect(rightPaddle.xpos, rightPaddle.ypos - 50, 25, 100);
}

function drawBall(ctx) {
    ctx.fillStyle = WHITE;
    ctx.fillRect(ball.xpos, ball.ypos, 10, 10);
}

function drawScore(ctx) {
    ctx.font = "60px Arial";
    ctx.strokeStyle = RED;
    ctx.strokeText(leftScore, 250, 50);
    ctx.strokeStyle = BLUE;
    ctx.strokeText(rightScore, 550, 50);
}

function drawFps(ctx) {
    ctx.font = "16px Arial";
    ctx.fillStyle = WHITE;
    ctx.fillText(`${fps.toFixed()} Frames/Second`, 650, 560);
}

function drawTimings(ctx) {
    ctx.font = "16px Arial";
    ctx.fillStyle = WHITE;
    const timing = timings.reduce((a, b) => { return a + b }, 0) / timings.length;
    ctx.fillText(`${timing.toFixed(2)} ms render`, 650, 580);
}

function draw() {
    const ctx = document.getElementById("canvasId").getContext("2d");
    drawBlackBackground(ctx);
    drawDottedLine(ctx);
    drawPaddles(ctx);
    drawBall(ctx);
    drawScore(ctx);
    drawFps(ctx);
    drawTimings(ctx);
}

function render() {
    const lastTiming = Date.now();
    calculateFps();
    update();
    draw();
    timings[totalFrames % timings.length] = Date.now() - lastTiming;
    this.setTimeout(render, 0);
}

File: 05_interacting_with_the_browser/pong.cpp
#include <emscripten/emscripten.h>
#include <emscripten/bind.h>

EM_JS(void, drawCanvas, (int width, int height), {
    const body = document.getElementsByTagName("body")[0];
    const canvas = document.createElement("canvas");
    canvas.id = "canvasId";
    canvas.width = width;
    canvas.height = height;
    body.appendChild(canvas);
});

enum class Move
{
    STATIONARY = 0,
    UP,
    DOWN
};

struct Paddle
{
    float xpos;
    float ypos;

    void moveUp()
    {
        if (ypos > 50)
        {
            ypos--;
        }
    }

    void moveDown()
    {
        if (ypos < 550)
        {
            ypos++;
        }
    }

    bool isAtPaddleLevel(float ball_ypos)
    {
        return (ball_ypos - 5 < ypos + 50) && (ball_ypos + 5 > ypos - 50);
    }
};

struct Ball
{
    float xpos = 395;
    float ypos = 295;
    float xspeed = 1;
    float yspeed = static_cast<float>(rand()) / static_cast<float>(RAND_MAX);

    bool isAtTopOrBottom()
    {
        return ypos - 5 < 0 || ypos + 5 > 595;
    }

    bool scoresOnRight()
    {
        return xpos > 775;
    }

    bool scoresOnLeft()
    {
        return xpos < 25;
    }

    bool doesHitPaddle(Paddle left, Paddle right)
    {
        return (xpos - 5 < 50 && left.isAtPaddleLevel(ypos)) || (xpos + 5 > 750 && right.isAtPaddleLevel(ypos));
    }

    void update()
    {
        xpos += xspeed;
        ypos += yspeed;
    }
};

struct GameState
{
    Ball ball{};
    Paddle left{25, 300};
    Paddle right{750, 300};
    Move move = Move::STATIONARY;
    int leftScore = 0;
    int rightScore = 0;
};

float calculateReflectionFactor(Ball ball, Paddle left, Paddle right)
{
    Paddle paddle = (ball.xpos > 400) ? right : left;
    const float factor = ball.ypos - paddle.ypos;
    return factor / 100;
}

void makeAIMove(Ball ball, Paddle &leftPaddle)
{
    int idealPosition = ball.ypos;
    if (ball.xspeed <= 0)
    {
        float turns = (ball.xpos - 50) / (-1 * ball.xspeed);
        idealPosition = ball.ypos + (ball.yspeed * turns);
    }

    if (idealPosition < leftPaddle.ypos)
    {
        leftPaddle.moveUp();
    }
    if (idealPosition > leftPaddle.ypos)
    {
        leftPaddle.moveDown();
    }
};

GameState createInitialGameState()
{
    drawCanvas(800, 600);
    return GameState{};
}

GameState updatePosition(GameState gameState)
{

    if (gameState.ball.xspeed == 0 && gameState.ball.yspeed == 0)
    {
        gameState.ball.xspeed = 1;
    }

    if (gameState.move == Move::UP)
    {
        gameState.right.moveUp();
    }
    if (gameState.move == Move::DOWN)
    {
        gameState.right.moveDown();
    }

    makeAIMove(gameState.ball, gameState.left);

    // ball touch on sides and paddles
    if (gameState.ball.isAtTopOrBottom())
    {
        gameState.ball.yspeed = -gameState.ball.yspeed;
    }
    if (gameState.ball.doesHitPaddle(gameState.left, gameState.right))
    {
        gameState.ball.xspeed = -gameState.ball.xspeed;
        gameState.ball.xspeed *= 1.05;
        gameState.ball.yspeed += calculateReflectionFactor(gameState.ball, gameState.left, gameState.right);
    }

    // create score
    if (gameState.ball.scoresOnRight())
    {
        gameState.ball = Ball{};
        gameState.leftScore += 1;
    }
    if (gameState.ball.scoresOnLeft())
    {
        gameState.ball = Ball{};
        gameState.rightScore += 1;
    }

    gameState.ball.update();

    return gameState;
}

EMSCRIPTEN_BINDINGS(gameState)
{

    // enum
    emscripten::enum_<Move>("Move")
        .value("STATIONARY", Move::STATIONARY)
        .value("UP", Move::UP)
        .value("DOWN", Move::DOWN);

    // struct
    emscripten::value_object<Ball>("Ball")
        .field("xpos", &Ball::xpos)
        .field("ypos", &Ball::ypos)
        .field("xspeed", &Ball::xspeed)
        .field("yspeed", &Ball::yspeed);

    emscripten::value_object<Paddle>("Paddle")
        .field("xpos", &Paddle::xpos)
        .field("ypos", &Paddle::ypos);

    emscripten::value_object<GameState>("GameState")
        .field("ball", &GameState::ball)
        .field("left", &GameState::left)
        .field("right", &GameState::right)
        .field("move", &GameState::move)
        .field("leftScore", &GameState::leftScore)
        .field("rightScore", &GameState::rightScore);

    // function
    emscripten::function("updatePosition", &updatePosition);
    emscripten::function("createInitialGameState", &createInitialGameState);
}
